docs/android-faqs.md

```diff
@@ -0,0 +1,119 @@
+
+There are a number of them. The issue boils down to the fact that you need to be
+completely certain that a JVM is associated with a given thread in order to call
+java code on it. The difficulty is that the JVM can GC its threads and will not
+let rust know about it. JNA can work around this for us to some extent, however
+there are difficulties.
+
+The approach it takes is essentially to spawn a thread for each callback
+invocation. If you are certain you’re going to do a lot of callbacks and they
+all originate on the same thread, you can tell it to cache these.
+
+Calling back from Rust into Kotlin isn’t too bad so long as you ensure the
+callback can not be GCed while rust code holds onto it, and you can either
+accept the overhead of extra threads being instantiated on each call, or you can
+ensure that it only happens from a single thread.
+
+Note that the situation would be somewhat better if we used JNI directly (and
+not JNA), but this would cause us to need to write two versions of each ffi
+crate, one for iOS, and one for Android.
+
+Ultimately, in any case where you can reasonably move to making something a
+blocking call, do so. It’s very easy to run such things on a background thread
+in Kotlin. This is in line with the Android documentation on JNI usage, and my
+own experience. It’s vastly simpler and less painful this way.
+
+(Of course, not every case is solvable like this).
+
+### Why are we using JNA rather than JNI, and what tradeoffs does that involve?
+
+We get a couple things from using JNA that we wouldn't with JNI.
+
+1. We are able to write a *single* FFI crate. If we used JNI we'd need to write
+   one FFI that android calls, and one that iOS calls.
+
+2. JNA provides a mapping of threads to callbacks for us, making callbacks over
+   the FFI possible. That said, in practice this is still error prone, and easy
+   to misuse/cause memory safety bugs, but it's required for cases like logging,
+   among others, and so it is a nontrivial piece of complexity we'd have to
+   reimplement.
+
+However, it comes with the following downsides:
+
+1. JNA has bugs. In particular, its not safe to use bools with them, it thinks
+   they are 32 bits, when on most platforms (every platform Rust supports) they
+   are 8 bits. They've been unwilling to fix the issue due to it breaking
+   backwards compatibility (which is... somewhat fair, there is a lot of C89
+   code out there that uses `bool` as a typedef for a 32-bit `int`).
+2. JNA makes it really easy to do the wrong thing and have it work but corrupt
+   memory. Several of the caveats around this are documented in the
+   [`ffi_support` docs](https://docs.rs/ffi-support/*/ffi_support/), but a
+   major one is when to use `Pointer` vs `String` (getting this wrong will
+   often work, but may corrupt memory).
+
+### How do I debug Rust code with the step-debugger in Android Studio
+
+1. Uncomment the `packagingOptions { doNotStrip "**/*.so" }` line from the
```

Comment:
TIL :)

Reply:
Haha, yes, I learned this about two days ago when copy-pasting this from one markdown document to another 😆